File: 01/python/main.py
from collections import Counter

def read_input_file(filename):
    """
    Reads a file containing pairs of integers and returns two sorted lists
    of integers (left and right).
    """

    left_list, right_list = [], []

    with open(filename, "r") as file:
        for line in file:
            left, right = list(map(int, line.split()))
            left_list.append(left)
            right_list.append(right)

    left_list.sort()
    right_list.sort()

    return left_list, right_list

def calculate_difference(left_list, right_list):
    """
    Calculate the sum of differences between corresponding elements of two
    sorted lists.
    """

    return sum(abs(left - right) for left, right in zip(left_list, right_list))

def calculate_similarity(left_list, right_list):
    """
    Calculates the similarity score by summing the product of elements in the
    left list with the number of its occurences in the right list.
    """

    right_counter = Counter(right_list)
    return sum(left * right_counter[left] for left in left_list)

def solve_part1(filename):
    """
    Solves Day 1, Part 1: calculates the total difference between the two
    sorted lists.
    """

    left_list, right_list = read_input_file(filename)
    return calculate_difference(left_list, right_list)

def solve_part2(filename):
    """
    Solves Day 1, Part2: calculates the total similarity score between two
    sorted lists.
    """

    left_list, right_list = read_input_file(filename)
    return calculate_similarity(left_list, right_list)

def main():
    """
    Main: reads input, solves the two parts, and prints the answers.
    """

    filename="../input"

    part1_answer = solve_part1(filename)
    print(f"Part 1: {part1_answer}")

    part2_answer = solve_part2(filename)
    print(f"Part 2: {part2_answer}")

if __name__ == "__main__":
    main()

File: 01/rust/src/main.rs
use std::collections::HashMap;
use std::fs::File;
use std::io::{self, BufReader, BufRead};

fn main() -> io::Result<()> {
    let filename = "../input";

    let (left_vec, right_vec) = read_input_file(filename)?;

    let part1_solution = calculate_difference(&left_vec, &right_vec);
    println!("Part 1: {}", part1_solution);

    let part2_solution = calculate_similarity(&left_vec, &right_vec);
    println!("Part 2: {}", part2_solution);

    Ok(())
}

fn read_input_file(filename: &str) -> io::Result<(Vec<i32>, Vec<i32>)> {
    let file = File::open(filename)?;
    let reader = BufReader::new(file);

    let mut left_vec = Vec::new();
    let mut right_vec = Vec::new();

    for line in reader.lines() {
        let line = line?;
        let mut parts = line.split_whitespace();

        if let (Some(left), Some(right)) = (parts.next(), parts.next()) {
            if let (Ok(left), Ok(right)) = (left.parse(), right.parse()) {
                left_vec.push(left);
                right_vec.push(right);
            } else {
                eprintln!("Warning: Invalid number format in line: {}", line);
            }
        }
    }

    left_vec.sort();
    right_vec.sort();

    Ok((left_vec, right_vec))
}

fn calculate_difference(left_vec: &[i32], right_vec: &[i32]) -> i32 {
    left_vec.iter()
        .zip(right_vec.iter())
        .map(|(left, right)| (left - right).abs())
        .sum()
}

fn calculate_similarity(left_vec: &[i32], right_vec: &[i32]) -> i32 {
    let mut counter = HashMap::new();
    for &num in right_vec {
        *counter.entry(num).or_insert(0) += 1;
    }

    left_vec.iter()
        .map(|&num| num * counter.get(&num).copied().unwrap_or(0) as i32)
        .sum()
}
